src/Raven.Server/Documents/Indexes/Static/TimeSeries/MapTimeSeriesIndex.cs

```diff
@@ -0,0 +1,219 @@
+
+        protected MapTimeSeriesIndex(MapIndexDefinition definition, StaticTimeSeriesIndexBase compiled)
+            : base(definition.IndexDefinition.Type, definition)
+        {
+            _compiled = compiled;
+
+            if (_compiled.ReferencedCollections == null)
+                return;
+
+            foreach (var collection in _compiled.ReferencedCollections)
+            {
+                foreach (var referencedCollection in collection.Value)
+                    _referencedCollections.Add(referencedCollection.Name);
+            }
+        }
+
+        protected override void SubscribeToChanges(DocumentDatabase documentDatabase)
+        {
+            if (_referencedCollections.Count > 0)
+                DocumentDatabase.Changes.OnDocumentChange += HandleDocumentChange;
+
+            if (DocumentDatabase != null)
+                DocumentDatabase.Changes.OnTimeSeriesChange += HandleTimeSeriesChange;
+        }
+
+        protected override void UnsubscribeFromChanges(DocumentDatabase documentDatabase)
+        {
+            if (_referencedCollections.Count > 0)
+                DocumentDatabase.Changes.OnDocumentChange -= HandleDocumentChange;
+
+            if (DocumentDatabase != null)
+                DocumentDatabase.Changes.OnTimeSeriesChange -= HandleTimeSeriesChange;
+        }
+
+        protected override void HandleDocumentChange(DocumentChange change)
+        {
+            if (_referencedCollections.Contains(change.CollectionName) == false)
+                return;
+
+            _mre.Set();
+        }
+
+        protected override IIndexingWork[] CreateIndexWorkExecutors()
+        {
+            var workers = new List<IIndexingWork>
+            {
+                //new CleanupDocuments(this, DocumentDatabase.DocumentsStorage, _indexStorage, Configuration, null)
+            };
+
+            if (_referencedCollections.Count > 0)
+                workers.Add(_handleReferences = new HandleTimeSeriesReferences(this, _compiled.ReferencedCollections, DocumentDatabase.DocumentsStorage.TimeSeriesStorage, DocumentDatabase.DocumentsStorage, _indexStorage, Configuration));
+
+            workers.Add(new MapTimeSeries(this, DocumentDatabase.DocumentsStorage.TimeSeriesStorage, _indexStorage, null, Configuration));
+
+            return workers.ToArray();
+        }
+
+        public override void HandleDelete(Tombstone tombstone, IIndexCollection collection, IndexWriteOperation writer, TransactionOperationContext indexContext, IndexingStatsScope stats)
+        {
+            throw new NotSupportedException();
+
+            //if (_referencedCollections.Count > 0)
+            //    _handleReferences.HandleDelete(tombstone, collection, writer, indexContext, stats);
+
+            //base.HandleDelete(tombstone, collection, writer, indexContext, stats);
+        }
+
+        public override long GetLastTombstoneEtagInCollection(DocumentsOperationContext databaseContext, IIndexCollection collection, bool isReference)
```

Comment:
What does `isReference` mean here?

Reply:
Need to check if we can do it better, but for TS we are not processing tombstones (because we do no have them), but this method is used to get also the last etag of referenced collections, for which we need to kip track. I've added that so I can distinguish properly if this is called from references or from 'regular' map. Probably it might end up as a flag inside the map, so the call be skipped in code, then no additional paramter will be needed.

Reply:
We still need to process tombstones of docs - if a document is deleted then we also need to handle it and delete related entries from index. I have introduced source document id (`sourceDocId()`) in order to handle that.